### PR TITLE
Implement classification metrics (#15)

### PR DESCRIPTION
**Pull request by Sarah5567**
## Description
Implemented evaluation setup based on research of classification metrics.  
Selected **2 metrics** for continuation:
- **Accuracy**
- **F1-score (micro)**

Implementation uses **scikit-learn** for simplicity and reliability.

## Changes
- Added `metrics/` package:
  - `api.py` – unified API (`compute_metrics`)
  - `sklearn_metrics.py` – backend with `scikit-learn`
  - `__init__.py` – exposes API
- Added `metrics_example.py`:
  - Generates toy multiclass dataset
  - Trains simple classifier (Logistic Regression)
  - Computes metrics
  - Prints results as a table with `pandas`

### CODE REVIEW

File: metrics/api.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+
+def compute_metrics(
+    y_true: Sequence[int],
+    y_pred: Sequence[int],
+    labels: Optional[Iterable[int]] = None,
+) -> Dict[str, float]:
+    """
+    Unified API: accepts ground truth and predictions (single-label multiclass).
+    Returns: {'accuracy': float, 'f1_micro': float}
+    """
```
Comment (Rut-Vahab) [inline metrics/api.py]: Docstring not accurate, should mention all returned metrics

File: metrics/api.py [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+    return {
+        "accuracy": accuracy_score_skl(y_true, y_pred),
+        "f1_micro": f1_micro_skl(y_true, y_pred, labels=labels),
+    }
```
Comment (Rut-Vahab) [inline metrics/api.py]: Only two metrics are returned, should allow extending or passing metrics as parameter

File: metrics/sklearn_metrics.py [hunk 3]
```diff
@@ -0,0 +1,15 @@
+from typing import Iterable, Optional, Sequence
+from sklearn.metrics import accuracy_score, f1_score
+
+
+def accuracy_score_skl(y_true: Sequence[int], y_pred: Sequence[int]) -> float:
+    return float(accuracy_score(y_true, y_pred))
+
```
Comment (Rut-Vahab) [inline metrics/sklearn_metrics.py]: Wrapper adds little value, could just use sklearn directly

File: metrics/sklearn_metrics.py [hunk 4]
```diff
@@ -0,0 +1,15 @@
...
+
+def f1_micro_skl(
+    y_true: Sequence[int],
+    y_pred: Sequence[int],
+    labels: Optional[Iterable[int]] = None,
+) -> float:
+    # micro averages TP/FP/FN over all classes; for single-label multiclass
+    return float(f1_score(y_true, y_pred, labels=labels, average="micro"))
```
Comment (Rut-Vahab) [inline metrics/sklearn_metrics.py]: Same here, seems redundant

**Comment by Rut-Vahab** [comment]
Function names with _skl suffix are unnecessary, could be simplified.

**Comment by Rut-Vahab** [comment]
metrics_example.py should be moved under examples/ instead of main code.

**Comment by chani0343** [comment]
## Missing Requirements from Original Task

Good work on the initial implementation! However, several key requirements from the original task (#15) still need to be addressed:

### 1. **Third Metric Required**
The task specifies "3 most important metrics" but only 2 are implemented (Accuracy + F1-micro). Based on our dataset analysis showing **extreme class imbalance** (39 classes, ranging from 1 to 1724 objects per class), we need to add **F1-macro** as the third metric.

F1-macro gives equal weight to all classes regardless of frequency, which is crucial for agricultural classification where rare crop types are still important to identify correctly.

### 2. **Missing Latency Measurement**
The task requires "Measures average latency with std" but there's no timing measurement. Please add:
```python
import time
import statistics
# Measure multiple runs and calculate mean/std of execution times

**Comment by lyuzinmaxim** [comment]
General comment: since you're calculating f1 score directly from sklearn, please also calculate explicitly precision and recall

File: metrics/sklearn_metrics.py (hunk 3 above)
Comment (lyuzinmaxim) [reply metrics/sklearn_metrics.py]: In general it's a nice way to "unify/standardize" the functions, but here since we're using the same library it's redundand

File: metrics/api.py [hunk 5]
```diff
@@ -0,0 +1,17 @@
+from typing import Dict, Iterable, Optional, Sequence
+from metrics.sklearn_metrics import accuracy_score_skl, f1_micro_skl
+
+
```
Comment (lyuzinmaxim) [inline metrics/api.py]: Please create a pydantic data model for storing the evaluation results. Just as example:
```
from pydantic import BaseModel, Field

class ClassificationReport(BaseModel):
    accuracy: float = Field(..., , ge=0.0, le=1.0, description="Overall accuracy across all classes")
    f1_score_micro: float <SMTH>
    ....
```

Then the return of `compute_metrics` would be of type `ClassificationReport` which is much more readable:
```
def compute_metrics(
    y_true: Sequence[int],
    y_pred: Sequence[int],
    labels: Optional[Iterable[int]] = None,
) -> ClassificationReport:
   ...
   return ClassificationReport(
        accuracy=accuracy,
        f1_score_macro=f1_macro,
    )

```


What we get if we're using pydantic as a data class:
- auto validation of data type (e.g., `float`), data range (see `ge`, `le`)
- easy conversion to json/dicts/another serialization
- easy to read & transparent
- easy to maintain (extend/inherit/change/...)

**Comment by lyuzinmaxim** [comment]
Please add a documentation explaining the metrics

File: env/requirements.txt [hunk 6]
```diff
@@ -15,6 +15,7 @@ matplotlib==3.8.4
 tqdm==4.66.2
 pillow==10.3.0
 requests==2.32.3
+pydantic>=2.6,<3
```
Comment (lyuzinmaxim) [inline env/requirements.txt]: please use the fixed version, i.e. `==`

File: metrics/api.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+        accuracy=acc,
+        precision_micro=prec_micro,
+        recall_micro=rec_micro,
+        f1_micro=f1_micro,
+        latency_mean=latency_mean,
+        latency_std=latency_std,
+    )
+
+
+def measure_latencies(
```
Comment (lyuzinmaxim) [inline metrics/api.py]: Sorry I am confused about how the latency is being computed. Could you please describe the flow?

File: metrics/metrics_example.py [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+
+    # 5) Calculate metrics with the unified API (Pydantic model)
+    report: ClassificationReport = compute_metrics(
+        y_true=y_test,
+        y_pred=y_pred,
+        latencies=latencies,
+    )
+
+    # 6) Summary results table (Pandas)
+    df = pd.DataFrame(
```
Comment (lyuzinmaxim) [inline metrics/metrics_example.py]: I think the conversion to pandas dataframe can be done more easily - please check the docs: https://docs.pydantic.dev/1.10/usage/exporting_models/

Comment (lyuzinmaxim) [reply metrics/metrics_example.py]: because now we're kinda not using all the features of pydantic and doing things manually.

File: metrics/api.py (hunk 7 above)
Comment (Sarah5567) [reply metrics/api.py]: The function first performs a few warmup calls to avoid measuring initialization overhead.  
Then, for each repetition, it records a start timestamp with `time.perf_counter()`, runs `predict_fn(inputs)`, records the end timestamp, and appends the duration (`t1 - t0`) to a list.

Comment (lyuzinmaxim) [reply metrics/api.py]: Could you please explain what's in each "repetition"? How can we use it for the model inference?

Comment (Sarah5567) [reply metrics/api.py]: Each repetition is one full inference call of `predict_fn(inputs)` measured separately.  
We record the start time, run the model once, record the end time, and store the duration.  
The list of repetitions lets us compute statistics (e.g., mean and std) to characterize model inference latency.

Comment (lyuzinmaxim) [reply metrics/api.py]: Are we gonna have the `predict_fn` in the real code for our classification models?

Comment (Sarah5567) [reply metrics/api.py]: Yes. we will pass to `predict_fn` the actual function that performs the prediction for our classification models.

Comment (lyuzinmaxim) [reply metrics/api.py]: Ok, I see now, thanks. what does "fn" mean in the function name?

Comment (Sarah5567) [reply metrics/api.py]: "fn" is just a shorthand for "function".

File: metrics/api.py (hunk 2 above)
Comment (lyuzinmaxim) [reply metrics/api.py]: Could you please rename this file since it's not reflecting the content.

File: metrics/api.py (hunk 7 above)
Comment (lyuzinmaxim) [reply metrics/api.py]: It's interfering with fn as "false negatives". Please use smth different.

Comment (Sarah5567) [reply metrics/api.py]: We changed it to "predict_func"

File: metrics/api.py (hunk 2 above)
Comment (Sarah5567) [reply metrics/api.py]: renamed to `metrics_api.py`, is that fine?

**Review by lyuzinmaxim — APPROVED** [review]
